### PR TITLE
Allow uploading attachment files alongside a submission

### LINKED ISSUE #102: Extend submission.create to support attachments
I sometimes have to pull data from a device and send it to a server. I'd like to script that with pyODK but there's no easy way to create a submission with attachments. It'd be nice to have that.

Here's what I'm doing...
```python
#!/usr/bin/env python3

from pathlib import Path
import requests

SUBMISSION_URL = "https://foo.getodk.cloud/v1/key/1234567890/projects/1/submission"
INSTANCES_PATH = Path("/Users/me/Desktop/instances")

def submit_instances(folder_path):
    files = {}

    for file in folder_path.iterdir():
        if file.is_file():
            file_content = file.read_bytes()
            
            if file.suffix == ".xml":
                print(file.name)
                files["xml_submission_file"] = ("instance.xml", file_content)
            else:
                files[file.name] = (file.name, file_content)

    response = requests.post(
        SUBMISSION_URL,
        files=files,
        headers={"X-OpenRosa-Version": "1.0"}
    )

    if response.status_code != 201:
        print(response.text)

for folder in INSTANCES_PATH.iterdir():
    if folder.is_dir():
        submit_instances(folder)
```

### PR DESCRIPTION
**Pull request by spwoodcock**
## This PR

- Closes #102
- Following on from https://github.com/getodk/pyodk/pull/104 (built on the branch, anticipating the merge)
- Allow passing a list of file paths in the SubmissionService.create() method.
- The files are uploaded as attachments after the submission is created.

If easier, https://github.com/getodk/pyodk/pull/103 and https://github.com/getodk/pyodk/pull/104 could be closed and this reviewed & merged all in one go.

#### What has been done to verify that this works as intended?

I added a test to upload 3 attachment files to one submission:

image

#### Why is this the best possible solution? Were any other approaches considered?

- It solves a requirement provided from an existing issue.
- The other approach would be to expect the user to create the submission, then upload attachments separately.
- I think this is a nice addition though, making life significantly easier to creating a submission+attachment combo.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

- This shouldn't really impact the existing API.
- It's additive only changes to the submission create.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.

- Updates to the pyodk docs!
- Will look into that if this PR is approved.

#### Before submitting this PR, please make sure you have:
- [x] included test cases for core behavior and edge cases in `tests`
- [x] run `python -m unittest` and verified all tests pass
- [x] run `ruff format pyodk tests` and `ruff check pyodk tests` to lint code
- [x] verified that any code or assets from external sources are properly credited in comments

### CODE REVIEW

File: pyodk/client.py [hunk 1]
```diff
@@ -66,6 +67,11 @@ def __init__(
         self.submissions: SubmissionService = SubmissionService(
             session=self.session, default_project_id=self.project_id
         )
+        self.submission_attachments: SubmissionAttachmentService = (
```
Comment (lindsay-stevens) [inline pyodk/client.py:70]: Please remove from client - the current use case is to upload submission + attachments at the same time, rather than separately.

File: pyodk/_endpoints/submissions.py [hunk 2]
```diff
@@ -201,6 +201,8 @@ def create(
         project_id: int | None = None,
         device_id: str | None = None,
         encoding: str = "utf-8",
+        # Here we must use imported typing.List to avoid conflict with .list method
+        attachments: builtins.list[str] | None = None,
```
Comment (lindsay-stevens) [inline pyodk/_endpoints/submissions.py:205]: Are you using an old python version? I think the `list` annotation is OK for all current versions. The previous equivalent was to using `typing.List`, not builtins. Anyway that's besides the point - we just need something to feed the `for` loop i.e. `Iterable[PathLike | str]`

File: pyodk/_endpoints/submission_attachments.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+                instance_id=iid,
+                fname=file_name,
+            ),
+            logger=log,
+        )
+        return response.content
+
+    def upload(
+        self,
+        file_path_or_bytes: PathLike | str | bytes,
```
Comment (lindsay-stevens) [inline pyodk/_endpoints/submission_attachments.py:122]: Please follow the `forms.create` process where it's assumed file paths are passed in. If using bytes is needed please open a ticket for that so we can think about doing that consistently across endpoints that currently accept file paths.

**Comment by lindsay-stevens** [comment]
Could you please rebase on the main branch? It seems like #104 has some of these changes here but not all of them so I'm not sure I understand the purpose of that PR. Also it seems like #103 was force pushed so I'll need to review again. Generally it will make it easier to review without all the model changes stuff.

Another thing needed before this can be merged is a real (E2E) test case. I've been putting those in `test_client.py` - it should be able to create what it needs to run, as long as it's a valid project ID and the user has privileges. Do you have an instance of Central that you can test with? Or do you have access to the ODK dev/staging instances? This is needed to verify that it works with current Central, sanity check against the original requirement (#102), and provide example code that use case. Sometimes the use cases are polished up into a module for `docs` - if you want to do that as well it would be appreciated.

**Comment by spwoodcock** [comment]
You told me the commit history had too many commits referencing formatting. So I squashed them into one commit per PR 😅

Is there a preferred way you would like commits to be made?

**Comment by spwoodcock** [comment]
> Could you please rebase on the main branch? It seems like #104 has some of these changes here but not all of them so I'm not sure I understand the purpose of that PR. Also it seems like #103 was force pushed so I'll need to review again. Generally it will make it easier to review without all the model changes stuff.
> 
> Another thing needed before this can be merged is a real (E2E) test case. I've been putting those in `test_client.py` - it should be able to create what it needs to run, as long as it's a valid project ID and the user has privileges. Do you have an instance of Central that you can test with? Or do you have access to the ODK dev/staging instances? This is needed to verify that it works with current Central, sanity check against the original requirement (#102), and provide example code that use case. Sometimes the use cases are polished up into a module for `docs` - if you want to do that as well it would be appreciated.

Sure, I can give that a go!

I have multiple ODK instances to test with, no worries. I generally test with a containerised Central.

I am using this code in production now & all seems to be well. Unfortunately I use pytest for all the test cases using this functionality, but I  can look at writing some stdlib unittest instead 👍

**Comment by spwoodcock** [comment]
Also, as mentioned in each PR, they anticipate the merge of the previous as they rely in code from the previous PR.

(1) Pydantic V2 models --> (2) Submission attachment service --> (3) Using the submission attachment service to upload attachments during submission

**Comment by lognaturel** [comment]
Thanks so much for taking this on, @spwoodcock, and making clear it's a priority for you. Looks like you saw @lindsay-stevens has picked it back up as https://github.com/getodk/pyodk/pull/110 so you don't have to take on additional testing and fiddly commit history concerns. We'll try to get it in soon!

**Comment by spwoodcock** [comment]
> Thanks so much for taking this on, @spwoodcock, and making clear it's a priority for you. Looks like you saw @lindsay-stevens has picked it back up as https://github.com/getodk/pyodk/pull/110 so you don't have to take on additional testing and fiddly commit history concerns. We'll try to get it in soon!

Hope this was somewhat useful 🙏

Not sure if my time is best spent contributing to pyodk though - it seems Lindsay is on top of things 👍

Looking forward to a new release
